### PR TITLE
HT5 ver 1

### PR DESCRIPTION
**Pull request by polisvv**

### CODE REVIEW

File: HT5.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    inp_location = pub.news_location()
+    pub.write_to_list_news()
+elif record_type_index == 2:
+    pub = Advertising(inp_name)
+    pub.write_to_list_ad()
+elif record_type_index == 3:
+    pub = Joke(inp_name)
+    pub.write_to_list_joke()
+else:
+    print('Something went wrong...')
```
Comment (romkabauer) [inline HT5.py:121]: Please, be more specific, like "Please, choose valid publication type 1 - News,2 - Private Ad or 3 - Smth"

File: HT5.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+record_type = [None, 'News', 'Advertising', 'Joke']  # None just for start from 1 index
+print('Select type of new record:')
+i = 1
+while i < len(record_type):
+    print(i, '-', record_type[i])
+    i = i + 1
+
+record_type_index = int(input(f'Select value between 1 and {len(record_type) - 1}: '))
+print('You selection is ', check_record_type(record_type_index))
+inp_name = input('Type publication name: ')
```
Comment (romkabauer) [inline HT5.py]: I think, it's better to validate record type before let user proceed with other inputs

File: HT5.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    return record_type[input_index]
+
+
+pub_list = []
+
+""" to input type of record """
+record_type = [None, 'News', 'Advertising', 'Joke']  # None just for start from 1 index
+print('Select type of new record:')
+i = 1
+while i < len(record_type):
```
Comment (romkabauer) [inline HT5.py]: `for object in iterable:` is more pythonic and readable way for cycles. It makes objects in the cycle visible and let name them properly also for readability

File: HT5.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        return pub_list
+
+
+class Advertising(Publication):
+    def __init__(self, name):
+        Publication.__init__(self, name=name)
+
+    def due_date(self):
+        print('Type advertising expiration date')
+        exp_year = int(input('Enter a year: '))
```
Comment (romkabauer) [inline HT5.py]: What if user input text? or random number like 999999? Users can be unique persons:) I think we need to handle it:)

File: HT5.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    def write_to_list_news(self):
+        pub_list.append(inp_location + ', ' + self.get_publish_date())
+        return pub_list
+
+
+class Advertising(Publication):
+    def __init__(self, name):
+        Publication.__init__(self, name=name)
+
+    def due_date(self):
```
Comment (romkabauer) [inline HT5.py]: please use verbs for function names, it will let other people know what exactly is the purpose of the function

File: HT5.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+class News(Publication):
+    def __init__(self, name, publish_datetime):
+        Publication.__init__(self, name=name)
+        self.publish_datetime = publish_datetime
+
+    def get_publish_date(self):
+        return self.publish_datetime.strftime('%Y-%m-%d %H:%M')
+
+    def news_location(self):
```
Comment (romkabauer) [inline HT5.py]: please use verbs for function names, it will let other people know what exactly is the purpose of the function

File: HT5.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    def write_to_list(self):
+        pub_list.insert(0, pub.name)
+        pub_list.insert(1, self.get_publication_text())
+        return pub_list
+
+    def write_to_file(self, l1):
+        file = open('myfile.txt', 'a+')
+        file.write('---------------------------------' + '\n')
+        file.write(l1[0] + '\n' + l1[1] + '\n' + l1[2] + '\n')
```
Comment (romkabauer) [inline HT5.py]: try to use f-string for concatenation
`f"{l1[0]}\nl1[1]..."`

File: HT5.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
+import datetime
+import random
+
+
+class Publication:
+    def __init__(self, name):
+        self.name = name
```
Comment (romkabauer) [inline HT5.py]: I think it is more reasonable to have parameters of the entity here, like "title", "body".
And then you can extend child classes with "specific_parameter" like expiration date or smth

File: HT5.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    def write_to_list_ad(self):
+        pub_list.append(self.due_date())
+        return pub_list
+
+
+class Joke(Publication):
+    def __init__(self, name):
+        Publication.__init__(self, name=name)
+
+    def random_funny_meter(self):
```
Comment (romkabauer) [inline HT5.py]: please use verbs for function names, it will let other people know what exactly is the purpose of the function

File: myfile.txt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+---------------------------------
+
+---------------------------------
+Private Ad
+I want to rent an apartments
+Actual until: 2022-11-01, 21 days left
+---------------------------------
+
+---------------------------------
+Weather forecast
```
Comment (romkabauer) [inline myfile.txt]: Don't see corresponding abstract class or certain class for this publication:)

File: HT5.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    if input_index == 0 or input_index >= len(record_type):
+        print('Invalid record type')
+        exit()
+    return record_type[input_index]
+
+
+pub_list = []
+
+""" to input type of record """
+record_type = [None, 'News', 'Advertising', 'Joke']  # None just for start from 1 index
```
Comment (romkabauer) [inline HT5.py]: I think there is no need to start from 1:) Just guide user what number corresponds to each type:)

**Review by romkabauer — CHANGES_REQUESTED** [review]

File: HT5.py (hunk 1 above)
Comment (polisvv) [reply HT5.py:121]: Discussed in teams.

File: HT5.py (hunk 2 above)
Comment (polisvv) [reply HT5.py]: Discussed in teams

File: HT5.py (hunk 3 above)
Comment (polisvv) [reply HT5.py]: changed

File: HT5.py (hunk 4 above)
Comment (polisvv) [reply HT5.py]: Python returns an error: ValueError: year 999999 is out of range. Do we need additional checks here?

File: HT5.py (hunk 5 above)
Comment (polisvv) [reply HT5.py]: ok

File: HT5.py (hunk 6 above)
Comment (polisvv) [reply HT5.py]: ok

File: HT5.py (hunk 7 above)
Comment (polisvv) [reply HT5.py]: Ok

File: HT5.py (hunk 8 above)
Comment (polisvv) [reply HT5.py]: Ok. Just want to emphasize that lector used "name" in the example and I can't read thoughts

File: HT5.py (hunk 9 above)
Comment (polisvv) [reply HT5.py]: ok

File: myfile.txt (hunk 10 above)
Comment (polisvv) [reply myfile.txt]: It is same as news, but another name. I'll delete possibility to insert random name by user

File: HT5.py (hunk 11 above)
Comment (polisvv) [reply HT5.py]: Ok, but not sure why I can't do it my way

File: HT5.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+class Publication:
+    def __init__(self, title):
+        self.title = title
+
+    def get_publication_text(self):
+        body = input('Type text: ')
+        return body
+
+    def write_to_list(self):
+        pub_list.insert(0, pub.title)
```
Comment (romkabauer) [inline HT5.py]: It's not a good idea to write smth into a global variable being in the context of the function. Let's require some list as a parameter to let function write into any list, user can provide

File: HT5.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        pub_list.insert(1, self.get_publication_text())
+        return pub_list
+
+    def write_to_file(self, l1):
```
Comment (romkabauer) [inline HT5.py]: Could we choose a more understandable name for the parameter?

File: HT5.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+class News(Publication):
+    def __init__(self, title, publish_datetime):
+        Publication.__init__(self, title=title)
+        self.publish_datetime = publish_datetime
+
+    def get_publish_date(self):
+        return self.publish_datetime.strftime('%Y-%m-%d %H:%M')
+
+    def get_news_location(self):
+        location = input('Type city name: ')
```
Comment (romkabauer) [inline HT5.py:34]: TO REVIEW:
What if input type will change? We will need to write another function to take data for example from API-call. Taking into account that we have quite many publication types, we will need to implement as many new functions as many publication types we have. It's an argument for abstraction of a class for inputting data

File: HT5.py (hunk 4 above)
Comment (romkabauer) [reply HT5.py]: Errors are causing script to fail, I'm suggesting to add the logic for handing these cases without script fails

**Review by romkabauer — CHANGES_REQUESTED** [review]

File: HT5.py (hunk 4 above)
Comment (polisvv) [reply HT5.py]: ok

File: HT5.py (hunk 13 above)
Comment (polisvv) [reply HT5.py]: ok

File: HT5.py (hunk 12 above)
Comment (polisvv) [reply HT5.py]: ok

Comment (polisvv) [reply HT5.py]: Can I use same name for variable inside a few function?

File: HT5.py (hunk 14 above)
Comment (polisvv) [reply HT5.py:34]: Could you please elaborate on that? I didn't get your point.

File: HT5.py (hunk 12 above)
Comment (romkabauer) [reply HT5.py]: of course 
More on that: 
- https://www.w3schools.com/python/python_scope.asp 
- https://www.datacamp.com/tutorial/scope-of-variables-python

Comment (polisvv) [reply HT5.py]: Thanks, done

File: HT5.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    print(num, record)
+
+record_type_index = int(input(f'Select value between 0 and {len(record_type) - 1}: '))
+print('You selection is ', check_record_type(record_type_index))
+
+
+if record_type_index == 0:
+    pub = News(record_type[record_type_index], datetime.datetime.now())
+    inp_location = pub.get_news_location()
+    pub.write_to_list_news()
```
Comment (romkabauer) [inline HT5.py]: argument missing, I believe there should be pub_list?

File: HT5.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+    def get_publish_date(self):
+        return self.publish_datetime.strftime('%Y-%m-%d %H:%M')
+
+    def get_news_location(self):
+        location = input('Type city name: ')
+        return location
+
+    def write_to_list_news(self, publication_list):
+        publication_list.append(f'{inp_location}, {self.get_publish_date()}')
```
Comment (romkabauer) [inline HT5.py]: Same notice like in previous comment, please take news location as function argument, not global variable

**Review by romkabauer — CHANGES_REQUESTED** [review]

File: HT5.py (hunk 15 above)
Comment (polisvv) [reply HT5.py]: yes

File: HT5.py (hunk 16 above)
Comment (polisvv) [reply HT5.py]: changed

**Review by romkabauer — APPROVED** [review]

**Pull request merged**